desc/particles.py

```diff
@@ -0,0 +1,1215 @@
+        magnetic moment (mv⊥²/2|B|) of each particle. The second output of
+        ``initializer.init_particles``.
+    stepsize_controller : diffrax.AbstractStepsizeController
+        Stepsize controller to use for the integration. If not provided, a
+        PIDController with the given ``rtol`` and ``atol`` will be used.
+    saveat : diffrax.SaveAt
+        SaveAt object to specify where to save the output. If not provided, will
+        save at the specified times in ts.
+    event : diffrax.Event
+        Custom event function to stop integration. If not provided, the default
+        event function will be used, which stops integration when particles leave the
+        bounds. If integration is stopped by the event, the output will contain NaN
+        values for the points outside the bounds.
+    """
+    # convert cartesian-like for integration in flux coordinates
+    if isinstance(field, Equilibrium):
+        xp = y0[:, 0] * jnp.cos(y0[:, 1])
+        yp = y0[:, 0] * jnp.sin(y0[:, 1])
+        y0 = y0.at[:, 0].set(xp)
+        y0 = y0.at[:, 1].set(yp)
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", message="unhashable type")
+        # we only want to map over initial positions and particle arguments
+        # Note: vmap with keyword arguments is weird, not using it for now
+        yt = vmap_chunked(
+            _intfun_wrapper, in_axes=(0, 0) + 12 * (None,), chunk_size=chunk_size
+        )(
+            y0,
+            model_args,
+            field,
+            params,
+            ts,
+            max_steps,
+            min_step_size,
+            solver,
+            stepsize_controller,
+            adjoint,
+            event,
+            model,
+            saveat,
+            options,
+        )
+
+    yt = jnp.where(jnp.isinf(yt), jnp.nan, yt)
+
+    x = yt[:, :, :3]
+    v = yt[:, :, 3:]
+
+    # convert back to flux coordinates
+    if isinstance(field, Equilibrium):
+        rho = jnp.sqrt(x[:, :, 0] ** 2 + x[:, :, 1] ** 2)
+        theta = jnp.arctan2(x[:, :, 1], x[:, :, 0])
+        x = x.at[:, :, 0].set(rho)
+        x = x.at[:, :, 1].set(theta)
+
+    return x, v
+
+
+def _intfun_wrapper(
+    x,
+    model_args,
+    field,
+    params,
+    ts,
+    max_steps,
+    min_step_size,
+    solver,
+    stepsize_controller,
+    adjoint,
+    event,
+    model,
+    saveat,
+    options,
+):
+    """Wrapper for the integration function for vectorized inputs.
+
+    Defining a lambda function inside the ``_trace_particles`` function leads
+    to recompilations, so instead we define the wrapper here.
+    """
+    return diffeqsolve(
+        terms=model,
+        solver=solver,
+        y0=x,
+        args=[model_args, field, params, options],
+        t0=ts[0],
+        t1=ts[-1],
+        saveat=saveat,
+        max_steps=max_steps,
+        dt0=min_step_size,
+        stepsize_controller=stepsize_controller,
+        adjoint=adjoint,
+        event=event,
+    ).ys
```

Comment:
we should have an option to return more of the diffrax.solution object, at least the `result` flag and `stats`, to check if the solution converged and if not why.

Reply:
my point is we probably don't want to be throwing errors at all, especially during optimization, it's usually preferable to just return some flag saying whether it succeeded or not and let the caller/objective etc handle failures (eg by ignoring those values).

I've found `stats` is useful for determining which solver and max_steps to use.